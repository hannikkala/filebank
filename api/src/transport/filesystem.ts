import fs = require("fs");
import path = require("path");
import fsExtra = require("fs-extra");
import glob from "glob";
import { promisify } from "util";
import {
  Directory,
  File,
  FileStorage,
  Item,
  ItemType,
  MoveDirectoryResponse
} from "../index";

const DEFAULT_MIMETYPE = "application/octet-stream";

const deleteFolderRecursive = (path: string) => {
  fs.readdirSync(path).forEach((file) => {
    const curPath = `${path}/${file}`;
    if (fs.statSync(curPath).isDirectory()) {
      // recurse
      deleteFolderRecursive(curPath);
    } else {
      // delete file
      fs.unlinkSync(curPath);
    }
  });
  fs.rmdirSync(path);
};

export class FilesystemStorage implements FileStorage {
  rootDir: string;
  constructor(rootDir: string) {
    this.rootDir = rootDir;

    if (!fs.existsSync(rootDir)) {
      fs.mkdirSync(rootDir);
    }
  }
  relatify(path: string): string {
    if (path.length >= this.rootDir.length) {
      return path.substring(this.rootDir.length + 1);
    }
    return path;
  }
  async list(id: string): Promise<(Directory | File)[]> {
    const dirPath: string = path.resolve(this.rootDir, id);
    const filenames: string[] = fs.readdirSync(dirPath);
    return filenames.map((filename) => {
      const ref = this.relatify(path.resolve(this.rootDir, id, filename));
      if (fs.statSync(path.resolve(dirPath, filename)).isDirectory()) {
        return {
          refId: ref,
          name: filename,
          type: ItemType.Directory
        };
      } else {
        return {
          refId: ref,
          name: filename,
          mimetype: DEFAULT_MIMETYPE,
          type: ItemType.File
        };
      }
    });
  }

  async mkdir(dirRef: string, dir: Directory): Promise<Directory> {
    const dirPath = path.resolve(this.rootDir, dirRef || "", dir.name);
    fs.mkdirSync(dirPath);
    return {
      refId: this.relatify(dirPath),
      name: dir.name,
      type: ItemType.Directory
    };
  }

  rmdir(dirId: string): Promise<void> {
    try {
      deleteFolderRecursive(path.resolve(this.rootDir, dirId));
      return Promise.resolve();
    } catch (e) {
      return Promise.reject(e);
    }
  }

  async getContent(file: File): Promise<Buffer> {
    return fs.readFileSync(path.resolve(this.rootDir, file.refId));
  }

  async removeFile(file: File): Promise<void> {
    fs.unlinkSync(path.resolve(this.rootDir, file.refId));
  }

  async createFile(
    file: File,
    directory: Directory | undefined,
    data: Buffer
  ): Promise<File> {
    const filePath = path.resolve(
      this.rootDir,
      directory ? directory.refId : "",
      file.name
    );
    fs.writeFileSync(filePath, data);
    return {
      mimetype: DEFAULT_MIMETYPE,
      refId: this.relatify(filePath),
      name: file.name,
      type: ItemType.File
    };
  }

  async moveFile(file: File, destination: Directory): Promise<File> {
    function move(
      oldPath: string,
      newPath: string,
      callback: (err?: any) => void
    ) {
      const copy = () => {
        const readStream = fs.createReadStream(oldPath);
        const writeStream = fs.createWriteStream(newPath);

        readStream.on("error", callback);
        writeStream.on("error", callback);

        readStream.on("close", () => {
          fs.unlink(oldPath, callback);
        });

        readStream.pipe(writeStream);
      };
      fs.rename(oldPath, newPath, (err) => {
        if (err) {
          if (err.code === "EXDEV") {
            copy();
          } else {
            callback(err);
          }
          return;
        }
        callback();
      });
    }

    const moveAsync = promisify(move);

    const oldPath = path.resolve(this.rootDir, file.refId);
    const filename = file.refId.split("/").pop();
    const destPath = path.resolve(
      this.rootDir,
      destination.refId,
      filename || ""
    );
    await moveAsync(oldPath, destPath);
    file.refId = this.relatify(destPath);
    return file;
  }

  async moveDirectory(
    dir: Directory,
    destination: Directory
  ): Promise<MoveDirectoryResponse> {
    const subdirRoot = path.resolve(this.rootDir, dir.refId);
    const globAsync = promisify(glob);
    const items = await globAsync("**", { cwd: subdirRoot });
    const destinationRef = destination.refId
      ? destination.refId
      : destination.name;
    const targetPath = fs.existsSync(path.resolve(this.rootDir, destinationRef))
      ? path.resolve(this.rootDir, destinationRef, path.basename(dir.refId))
      : path.resolve(this.rootDir, destinationRef);

    fsExtra.moveSync(subdirRoot, targetPath);
    const allItems: {
      oldItem: Item;
      newItem: Item;
    }[] = items.map((item: string) => {
      const stat = fs.statSync(path.resolve(targetPath, item));
      const oldItem = {
        name: path.basename(item),
        refId: path.join(dir.refId, item),
        type: stat.isDirectory() ? ItemType.Directory : ItemType.File
      };
      const newItem = {
        name: path.basename(item),
        refId: path.join(destinationRef, item),
        type: stat.isDirectory() ? ItemType.Directory : ItemType.File
      };
      return {
        oldItem,
        newItem
      };
    });
    return {
      directory: {
        type: ItemType.Directory,
        name: path.basename(targetPath),
        refId: this.relatify(targetPath)
      },
      items: allItems
    };
  }
}
