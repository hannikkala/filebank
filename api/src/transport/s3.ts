import { Directory, FileStorage, File, Item, ItemType } from '../index';
import Promise = require('bluebird');
import AWS = require('aws-sdk');
import S3 = require('aws-sdk/clients/s3');
import { DeleteObjectsRequest, GetObjectOutput, ListObjectsOutput } from 'aws-sdk/clients/s3';
import * as _ from 'lodash';
import path = require('path');
import { S3StorageOptions } from '../config';
import { EventEmitter } from 'events';

AWS.config.setPromisesDependency(Promise);
AWS.config.update({ region: 'eu-west-1' });

export class S3Storage extends EventEmitter implements FileStorage {
  _bucket: string;
  private _s3: AWS.S3;

  constructor(options: S3StorageOptions) {
    super();
    this._s3 = new AWS.S3(Object.assign({ apiVersion: '2006-03-01' }, options.clientOptions));
    this._bucket = options.bucket;
    this.initialize();
  }

  initialize(): Promise<any> {
    return Promise.fromNode(cb => this._s3.headBucket({ Bucket: this._bucket }, cb))
      .then(() => {
        this.emit('initialized');
      })
      .catch(() => {
        return this._s3.createBucket({
          Bucket: this._bucket,
          ACL: 'authenticated-read'
        }).promise()
          .then(() => {
            this.emit('initialized');
          });
      });
  }

  list(id: string): Promise<(Directory | File)[]> {
    return Promise.fromNode(cb => this._s3.listObjects({ Bucket: this._bucket }, cb))
      .then((data) => {
        return data.Contents.map((obj: S3.Object) => {
          return {
            type: obj.Key!.endsWith('/') ? ItemType.Directory : ItemType.File,
            name: obj.Key,
          };
        });
      });
  }

  mkdir(dirRef: string, dir: Directory): Promise<Directory> {
    const key = `${dirRef}${dir.name}/`;
    return Promise.fromNode(cb => this._s3.putObject({ Bucket: this._bucket, Key: key, ACL: 'authenticated-read' }, cb))
      .then(() => {
        return { name: dir.name, type: ItemType.Directory, refId: key };
      });
  }

  rmdir(dirId: string): Promise<void> {
    return Promise.fromNode(cb => this._s3.deleteObject({ Bucket: this._bucket, Key: dirId }, cb));
  }

  getContent(file: File): Promise<Buffer> {
    return Promise.fromNode(cb => this._s3.getObject({ Bucket: this._bucket, Key: file.refId }, cb))
      .then((data: GetObjectOutput) => {
        return _.isString(data.Body!) ? new Buffer(data.Body! as string) : data.Body! as Buffer;
      });
  }

  removeFile(file: File): Promise<void> {
    return Promise.fromNode(cb => this._s3.deleteObject({ Bucket: this._bucket, Key: file.refId }, cb));
  }

  createFile(file: File, directory: Directory | any, data: Buffer): Promise<File> {
    const key = `${directory ? directory.refId : ''}${file.name}`;
    return Promise.fromNode(cb => this._s3.upload({
      Bucket: this._bucket, Key: key, ACL: 'authenticated-read', Body: data
    }, cb))
      .then(() => {
        return { name: file.name, type: ItemType.File, refId: key, mimetype: file.mimetype };
      });
  }

  moveFile(item: File, destination: Directory): Promise<File> {
    const key = `${destination ? destination.refId : ''}${path.basename(item.refId)}`;
    return Promise.fromNode(cb => this._s3.copyObject({
      Bucket: this._bucket,
      CopySource: `${this._bucket}/${item.refId}`,
      Key: key
    }, cb))
      .then(() => {
        return Promise.fromNode(cb => this._s3.deleteObject({ Bucket: this._bucket, Key: item.refId }, cb));
      }).then(() => {
        return { name: item.name, type: ItemType.File, refId: key, mimetype: item.mimetype };
      });
  }

  moveDirectory(dir: Directory, destination: Directory, refCb: (oldItem: Item, newItem: Item) => Promise<void>): Promise<Directory> {
    const targetRef = destination.refId ? destination.refId : `${destination.name}/`;
    return Promise.all([
      Promise.fromNode(cb => this._s3.listObjects({ Bucket: this._bucket, Prefix: dir.refId }, cb)),
      Promise.fromNode(cb => this._s3.listObjects({ Bucket: this._bucket, Prefix: targetRef }, cb))
    ]).spread((data: ListObjectsOutput, exists: ListObjectsOutput) => {
        const deleteParams: DeleteObjectsRequest = {
          Bucket: this._bucket,
          Delete: {
            Objects: [],
          },
        };
        const targetPrefix = exists.Contents!.length ? `${targetRef}${dir.name}/` : `${targetRef}`;
        console.log('targetprefix', targetPrefix);
        return Promise.map(data.Contents!, (s3item: S3.Object) => {
          deleteParams.Delete.Objects.push({ Key: s3item.Key! });
          const newKey = s3item.Key!.replace(dir.refId, targetPrefix);
          console.log('copyobject', s3item.Key, newKey);
          return Promise.fromNode(cb => this._s3.copyObject({
            Bucket: this._bucket,
            CopySource: `${this._bucket}/${s3item.Key}`,
            Key: newKey,
          }, cb))
            .then(() => {
              console.log('refCb', s3item.Key);
              if (s3item.Key === dir.refId) return;
              return refCb({
                refId: s3item.Key!,
                name: path.basename(s3item.Key!),
                type: s3item.Key!.endsWith('/') ? ItemType.Directory : ItemType.File,
              }, {
                refId: newKey,
                name: path.basename(s3item.Key!),
                type: s3item.Key!.endsWith('/') ? ItemType.Directory : ItemType.File,
              });
            });
        }).then(() => {
          console.log('deleteParams');
          return Promise.fromNode(cb => this._s3.deleteObjects(deleteParams, cb));
        }).then(() => {
          return { name: path.basename(targetPrefix), type: ItemType.Directory, refId: targetPrefix };
        });
      })
  }

}

