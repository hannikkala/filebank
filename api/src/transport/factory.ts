import { FileStorage } from '../index';
import config from '../config/config';
import { FilesystemStorage } from './filesystem';
import { S3Storage } from "./s3";
import { FilesystemStorageOptions } from "../config";

export let storage: FileStorage;
const storageConfig = config.get("storage");

export default {
  getInstance: (): FileStorage => {
    if (storage) return storage;
    switch (storageConfig.enabled) {
      case 'filesystem':
        storage = new FilesystemStorage((storageConfig.filesystem as FilesystemStorageOptions).rootDir);
        return storage;
      case 's3':
        storage = new S3Storage(storageConfig.s3);
        return storage;
    }
    throw new Error(`Factory method failed. No require for storage ${storageConfig.enabled}`);
  },
};
