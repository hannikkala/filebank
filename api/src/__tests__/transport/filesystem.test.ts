import { ItemType } from "../../index";
import * as fs from "fs";
import { FilesystemStorage } from "../../transport/filesystem";

import * as path from "path";

import * as helpers from "../helpers";

const filesystemstorageroot = "/tmp/filebank";
process.env.FILESYSTEM_STORAGE_ROOT = filesystemstorageroot;
const filesystem = new FilesystemStorage(filesystemstorageroot);

beforeAll(async () => {
  helpers.deleteFolderRecursive(filesystemstorageroot);
  fs.mkdirSync(filesystemstorageroot);
});

afterAll(async () => {
  helpers.deleteFolderRecursive(filesystemstorageroot);
});

describe("Filesystem storage", () => {
  it("can create directory", async () => {
    const dir = await filesystem.mkdir("", {
      refId: "temp",
      name: "temp",
      type: ItemType.Directory
    });
    expect(dir).toBeTruthy();
    expect(dir.type).toEqual("directory");
    expect(
      fs.existsSync(path.resolve(filesystemstorageroot, "temp"))
    ).toBeTruthy();
  });

  it("returns Promise when error", async () => {
    try {
      await filesystem.mkdir("nonexisting", {
        refId: "temp",
        name: "temp",
        type: ItemType.Directory
      });
      fail(new Error("Should fail."));
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });

  it("can create subdirectory", async () => {
    const d = await filesystem.mkdir("", {
      refId: "subdir",
      name: "subdir",
      type: ItemType.Directory
    });
    const subdir = await filesystem.mkdir(d.refId, {
      refId: "temp",
      name: "temp",
      type: ItemType.Directory
    });
    expect(subdir.refId).toEqual("subdir/temp");
    expect(
      fs.existsSync(path.resolve(filesystemstorageroot, "subdir", "temp"))
    ).toBeTruthy();
  });

  it("can create file on root", async () => {
    const file = await filesystem.createFile(
      { name: "test.txt", refId: "test.txt", type: ItemType.File },
      undefined,
      new Buffer("test text")
    );
    expect(file).toBeTruthy();
    expect(file.type).toEqual(ItemType.File);
    expect(
      fs.existsSync(path.resolve(filesystemstorageroot, file.refId))
    ).toBeTruthy();
  });

  it("returns Promise when creating file failed", async () => {
    try {
      await filesystem.createFile(
        { name: "test.txt", refId: "test.txt", type: ItemType.File },
        { refId: "nonexist", name: "nonexist", type: ItemType.Directory },
        new Buffer("test text")
      );
      fail(new Error("Should fail."));
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });

  it("can create file on subdir", async () => {
    const dir = await filesystem.mkdir("", {
      refId: "subdir2",
      name: "subdir2",
      type: ItemType.Directory
    });
    const file = await filesystem.createFile(
      { name: "test.txt", refId: "test.txt", type: ItemType.File },
      dir,
      new Buffer("test text")
    );
    expect(file).toBeTruthy();
    expect(file.refId).toEqual("subdir2/test.txt");
    expect(
      fs.existsSync(path.resolve(filesystemstorageroot, file.refId))
    ).toBeTruthy();
  });

  it("can remove directory", async () => {
    const dir = await filesystem.mkdir("", {
      refId: "rmdir",
      name: "rmdir",
      type: ItemType.Directory
    });
    const file = await filesystem.createFile(
      { refId: "test.txt", name: "test.txt", type: ItemType.File },
      dir,
      new Buffer("testtesttest")
    );
    expect(
      fs.existsSync(path.resolve(filesystemstorageroot, file.refId))
    ).toBeTruthy();
    await filesystem.rmdir("rmdir");
    expect(
      fs.existsSync(path.resolve(filesystemstorageroot, "rmdir"))
    ).toBeFalsy();
  });

  it("returns Promise when removing directory fails", async () => {
    try {
      await filesystem.rmdir("nonexist");
      fail(new Error("Should fail."));
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });

  it("can get file content", async () => {
    const file = await filesystem.createFile(
      { refId: "content.txt", name: "content.txt", type: ItemType.File },
      undefined,
      new Buffer("testtesttest")
    );
    const buffer = await filesystem.getContent(file);
    expect(buffer.toString()).toEqual("testtesttest");
  });

  it("can list directory contents", async () => {
    const root = await filesystem.mkdir("", {
      refId: "content",
      name: "content",
      type: ItemType.Directory
    });
    await filesystem.createFile(
      { refId: "fileone.txt", name: "fileone.txt", type: ItemType.File },
      root,
      new Buffer("test")
    );
    await filesystem.createFile(
      { refId: "filetwo.txt", name: "filetwo.txt", type: ItemType.File },
      root,
      new Buffer("test2")
    );
    await filesystem.mkdir(root.refId, {
      refId: "subdir",
      name: "subdir",
      type: ItemType.Directory
    });
    const list = await filesystem.list(root.refId);
    expect(list).toHaveLength(3);
    expect(list.filter((item) => item.type === "file")).toHaveLength(2);
    expect(list.filter((item) => item.type === "directory")).toHaveLength(1);
  });

  it("can remove file", async () => {
    const file = await filesystem.createFile(
      { refId: "rmfile.txt", name: "rmfile.txt", type: ItemType.File },
      undefined,
      new Buffer("testtesttest")
    );
    await filesystem.removeFile(file);
    expect(
      fs.existsSync(path.resolve(filesystemstorageroot, "rmfile.txt"))
    ).toBeFalsy();
  });

  it("can move file", async () => {
    const file = await filesystem.createFile(
      { refId: "mvfile.txt", name: "mvfile.txt", type: ItemType.File },
      undefined,
      new Buffer("testtesttest")
    );
    const dir = await filesystem.mkdir("", {
      name: "moveto",
      refId: "moveto",
      type: ItemType.Directory
    });
    const newFile = await filesystem.moveFile(file, dir);
    expect(newFile.refId).toEqual("moveto/mvfile.txt");
    expect(
      fs.existsSync(path.resolve(filesystemstorageroot, "moveto", "mvfile.txt"))
    ).toBeTruthy();
  });

  it("can move directory and contents", async () => {
    const from = await filesystem.mkdir("", {
      name: "dirmovefrom",
      refId: "dirmovefrom",
      type: ItemType.Directory
    });
    await filesystem.mkdir("dirmovefrom", {
      name: "sub1",
      refId: "dirmovefrom/sub1",
      type: ItemType.Directory
    });
    await filesystem.createFile(
      {
        refId: "mvfile.txt",
        name: "dirmovefrom/mvfile.txt",
        type: ItemType.File
      },
      undefined,
      new Buffer("testtesttest")
    );
    await filesystem.mkdir("", {
      name: "dirmoveto",
      refId: "dirmoveto",
      type: ItemType.Directory
    });
    await filesystem.moveDirectory(from, {
      name: "dirmovefrom",
      refId: "dirmoveto/dirmovefrom",
      type: ItemType.Directory
    });
  });
});
