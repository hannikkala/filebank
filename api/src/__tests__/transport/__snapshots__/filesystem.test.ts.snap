// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filesystem storage returns Promise when creating file failed 1`] = `[Error: ENOENT: no such file or directory, open '/tmp/filebank/nonexist/test.txt']`;

exports[`Filesystem storage returns Promise when error 1`] = `[Error: ENOENT: no such file or directory, mkdir '/tmp/filebank/nonexisting/temp']`;

exports[`Filesystem storage returns Promise when removing directory fails 1`] = `[Error: ENOENT: no such file or directory, scandir '/tmp/filebank/nonexist']`;
